plugin/pkg/kubectl/client/plugin_client.go

```diff
@@ -0,0 +1,192 @@
+// based on an existing configuration
+func NewClient() (*PluginClient, error) {
+	// resolve kubeconfig location, prioritizing the --config global flag,
+	// then the value of the KUBECONFIG env var (if any), and defaulting
+	// to ~/.kube/config as a last resort.
+	home := os.Getenv("HOME")
+	kubeconfig := home + "/.kube/config"
+
+	kubeconfigEnv := os.Getenv("KUBECONFIG")
+	if len(kubeconfigEnv) > 0 {
+		kubeconfig = kubeconfigEnv
+	}
+
+	configFile := os.Getenv("KUBECTL_PLUGINS_GLOBAL_FLAG_CONFIG")
+	kubeConfigFile := os.Getenv("KUBECTL_PLUGINS_GLOBAL_FLAG_KUBECONFIG")
+	if len(configFile) > 0 {
+		kubeconfig = configFile
+	} else if len(kubeConfigFile) > 0 {
+		kubeconfig = kubeConfigFile
+	}
+
+	if len(kubeconfig) == 0 {
+		return nil, fmt.Errorf("error iniializing client. The KUBECONFIG environment variable must be defined")
+	}
+
+	clientConfig, _, err := clientFromConfig(kubeconfig)
+	if err != nil {
+		return nil, fmt.Errorf("error obtaining client configuration: %v", err)
+	}
+
+	err = applyGlobalOptionsToConfig(clientConfig)
+	if err != nil {
+		return nil, fmt.Errorf("error processing global plugin options: %v", err)
+	}
+
+	c, err := clientset.NewForConfig(clientConfig)
+	if err != nil {
+		return nil, fmt.Errorf("error obtaining a client from existing configuration: %v", err)
+	}
+
+	pluginClient := PluginClient{c, clientConfig}
+	return &pluginClient, nil
+}
+
+func clientFromConfig(path string) (*restclient.Config, string, error) {
+	if path == "-" {
+		cfg, err := restclient.InClusterConfig()
+		if err != nil {
+			return nil, "", fmt.Errorf("cluster config not available: %v", err)
+		}
+		return cfg, "", nil
+	}
+
+	rules := &clientcmd.ClientConfigLoadingRules{ExplicitPath: path}
+	credentials, err := rules.Load()
+	if err != nil {
+		return nil, "", fmt.Errorf("the provided credentials %q could not be loaded: %v", path, err)
+	}
+
+	cfg := clientcmd.NewDefaultClientConfig(*credentials, &clientcmd.ConfigOverrides{})
+	config, err := cfg.ClientConfig()
+	if err != nil {
+		return nil, "", fmt.Errorf("the provided credentials %q could not be used: %v", path, err)
+	}
+
+	namespace, _, _ := cfg.Namespace()
+	return config, namespace, nil
+}
+
+func applyGlobalOptionsToConfig(config *restclient.Config) error {
+	// impersonation config
+	impersonateUser := os.Getenv("KUBECTL_PLUGINS_GLOBAL_FLAG_AS")
+	if len(impersonateUser) > 0 {
+		config.Impersonate.UserName = impersonateUser
+	}
+
+	impersonateGroupRaw := os.Getenv("KUBECTL_PLUGINS_GLOBAL_FLAG_AS_GROUP")
+	if len(impersonateGroupRaw) > 0 {
+		impersonateGroup := []string{}
+		err := json.Unmarshal([]byte(impersonateGroupRaw), &impersonateGroup)
+		if err != nil {
+			return fmt.Errorf("error parsing global option %q: %v", "--as-group", err)
+		}
+		if len(impersonateGroup) > 0 {
+			config.Impersonate.Groups = impersonateGroup
+		}
+	}
+	// tls config
+
+	caFile := os.Getenv("KUBECTL_PLUGINS_GLOBAL_FLAG_CERTIFICATE_AUTHORITY")
```

Comment:
I think I may need more education because (like the impersonation stuff), it seems odd to me that plugins have their own mechanism that appear to be duplicating what kubectl does - like this CA stuff. Why isn't the plugin logic just using the same data that kubectl uses?

Reply:
To this and the above - I'm not 100% sure. These flags appear to be something passed from the kubectl plugin framework if it wants to execute a plugin as a different user, or in a different context than the calling envrionment. As to why someone would want to do that - I don't know.